### PR TITLE
[wait-for] Map queries

### PR DESCRIPTION
**Pull request by SimonRichardson**
This is the initial idea for adding map queries. The code introduces a
new set of tokens (left and right bracket) and we parse the index ast as
a infix operator. From there the query evaluator can just loop through
until we get to the bottom of the ast and collapse up.

The following are pefectly plausable:

    config["value"] == 1
    config["values"]["nested"] == 1

Additionally we now have function calling for global builtin methods, so
it's possible to do:

    len(config) > 0 && config["value"] == 1

## QA Steps

Model config queries:

```sh
juju bootstrap lxd test
juju wait-for model -m controller controller --query='len(config) > 0 && config["agent-stream"] == "released"'
```

Additionally, you can use print to dump out the config so it makes it easier
to see what the value of an identifier is. The `print` function always returns
what was passed in, because queries are based on expressions, so you can
always place print anywhere in a query as a no-op.

```sh
juju wait-for model -m controller controller --query='len(print(config)) > 0 && config["agent-stream"] == "released"'
```

### CODE REVIEW

**Review by manadart — COMMENTED** [review]
Code is good; QA pending.

File: cmd/plugins/juju-wait-for/application.go [hunk 1]
```diff
@@ -111,10 +111,14 @@ func (c *applicationCommand) waitFor(name string, deltas []params.Delta, q query
 		case *params.ApplicationInfo:
 			if entityInfo.Name == name {
 				scope := MakeApplicationScope(entityInfo)
-				if res, err := q.Run(scope); query.IsInvalidIdentifierErr(err) {
+				if res, err := q.BuiltinsRun(scope); query.IsInvalidIdentifierErr(err) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/plugins/juju-wait-for/application.go]: I agree, this is done now.

**Review by manadart — APPROVED** [review]
QA is good.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**